File: TestWAT/Controls/ToggleButton.vb
Imports System.ComponentModel

Public Class ToggleButton
	Inherits Button
	Private _toggled As Boolean = False

	Private TextOff As String
	Private ImageOff As Image

	<Localizable(True), Category("Appearance")>
	Public Property ImageOn As Image

	<Localizable(True), Category("Appearance")>
	Public Property TextOn As String

	<Browsable(True), Category("Behavior"), DefaultValue(False)>
	Public Property Toggled As Boolean
		Get
			Return _toggled
		End Get
		Set(value As Boolean)
			If value <> _toggled Then
				If value Then ToggleOn() Else ToggleOff()
			End If
		End Set
	End Property

	Protected Overrides Sub OnClick(e As EventArgs)
		Toggled = Not Toggled
		MyBase.OnClick(e)
		If Toggled Then OnToggleOn(e) Else OnToggleOff(e)
	End Sub

	Private Sub ToggleOn()
		_toggled = True
		TextOff = Text
		ImageOff = Image
		Text = TextOn
		Image = ImageOn
		'OnToggleOn(EventArgs.Empty)
	End Sub

	Private Sub ToggleOff()
		_toggled = False
		Text = TextOff
		Image = ImageOff
		'OnToggleOff(EventArgs.Empty)
	End Sub

#Region "Events"

	Protected Sub OnToggleOn(e As EventArgs)
		RaiseEvent ToggledOn(Me, e)
	End Sub

	Protected Sub OnToggleOff(e As EventArgs)
		RaiseEvent ToggledOff(Me, e)
	End Sub

	<Browsable(True), Category("Action")>
	Public Event ToggledOn(sender As Object, e As EventArgs)

	<Browsable(True), Category("Action")>
	Public Event ToggledOff(sender As Object, e As EventArgs)

#End Region

End Class

File: JANL.WinForms/ExtendedControls/ToggleButton.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;

namespace JANL
{
    public class ToggleButton : Button
    {
        private Image ImageOff;
        private string TextOff;

        #region Properties

        #region Designer

        private bool _toggled;

        [Localizable(true), Category("ToggleButton")]
        public Image ImageOn { get; set; }

        [Localizable(true), Category("ToggleButton")]
        public string TextOn { get; set; }

        [Browsable(true), Category("ToggleButton"), DefaultValue(false)]
        public bool Toggled
        {
            get => _toggled;
            set
            {
                if (value != _toggled)
                {
                    if (value) { ToggleOn(); }
                    else { ToggleOff(); }
                }
            }
        }

        #endregion Designer

        #endregion Properties

        protected override void OnClick(EventArgs e)
        {
            Toggled = !Toggled;
            base.OnClick(e);
            if (Toggled) { OnToggleOn(e); }
            else { OnToggleOff(e); }
        }

        private void ToggleOff()
        {
            _toggled = false;
            Text = TextOff;
            Image = ImageOff;
        }

        private void ToggleOn()
        {
            _toggled = true;
            TextOff = Text;
            ImageOff = Image;
            Text = TextOn;
            Image = ImageOn;
        }

        #region Events

        protected void OnToggleOff(EventArgs e) => ToggledOff?.Invoke(this, e);

        protected void OnToggleOn(EventArgs e) => ToggledOn?.Invoke(this, e);

        [Browsable(true), Category("Action")]
        public event EventHandler ToggledOff;

        [Browsable(true), Category("Action")]
        public event EventHandler ToggledOn;

        #endregion Events
    }
}